Create mesh for CSGObject to plot Sample Shapes

DanielMurphy22 opened the pull request:
**Description of work.**

A mesh can now be created for CSGObjects  to plot Sample Shapes, such as spheres + cylinders and shapes can be merged.

Refs #30607 

**To test:**

- Run the script below
- Edit `SetSample(...)` to use the xml strings in the script for different shapes, and check that sphere, cylinder and a sphere with a cylindrical hole in work!
- Feel free to try some weird and wonderful shapes and combinations of shapes: https://docs.mantidproject.org/nightly/concepts/HowToDefineGeometricShape.html#howtodefinegeometricshape


```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from mantid.api import AnalysisDataService as ADS


# A fake host workspace, replace this with your real one.
ws = CreateSampleWorkspace()

cuboid_xml = " \
<cuboid id='some-cuboid'> \
  <height val='20.0'  /> \
  <width val='50.0' />  \
  <depth  val='20.0' />  \
  <centre x='100.0' y='100.0' z='100.0'  />  \
  <rotation x='60' y='90' z='217' /> \
</cuboid>  \
<algebra val='some-cuboid' /> \
"

sphere_xml = " \
<sphere id='some-sphere'> \
    <centre x='0.0'  y='0.0' z='0.0' /> \
    <radius val='0.5' /> \
</sphere> \
<algebra val='some-sphere' /> \
"

cylinder_xml = " \
<cylinder id='some-cylinder'> \
  <centre-of-bottom-base r='0.0' t='0.0' p='0.0' />  \
  <axis x='0.0' y='0.2' z='0' /> \
  <radius val='1' /> \
  <height val='10.2' /> \
</cylinder> \
<algebra val='some-cylinder' /> \
"

merge_xml = ' \
<cylinder id="stick"> \
  <centre-of-bottom-base x="-0.5" y="0.0" z="0.0" /> \
  <axis x="1.0" y="0.0" z="0.0" /> \
  <radius val="0.2" /> \
  <height val="1.0" /> \
</cylinder> \
\
<sphere id="some-sphere"> \
  <centre x="0.0"  y="0.0" z="0.0" /> \
  <radius val="0.5" /> \
</sphere> \
\
<algebra val="some-sphere (#stick)" /> \
'

flat_circle = " \
<cylinder id=\"pixel-shape\"> \
  <centre-of-bottom-base x=\"0\" y=\"-0.0001\" z=\"0\"/> \
  <axis x=\"0\" y=\"1\" z=\"0\"/><radius val=\"0.004\"/> \
  <height val=\"0.0002\"/> \
</cylinder>"

# Set sample geometry of workspace to this CSG object
SetSample(ws, Geometry={'Shape': 'CSG', 'Value': cuboid_xml})

sample = ws.sample()
shape = sample.getShape()
mesh = shape.getMesh()

facecolors = ['royalblue','b','red','firebrick', 'gold', 'orange','green', 'darkgreen','grey','black', 'purple', 'mediumorchid']

mesh_polygon_a = Poly3DCollection(mesh, facecolors = facecolors, edgecolors='green', linewidths=0.1, alpha = 0.5)
'''Problem based on order of being added!'''

fig, axes = plt.subplots(subplot_kw={'projection':'mantid3d'})
axes.add_collection3d(mesh_polygon_a)

# Auto scale to the mesh size
axes_lims = mesh.flatten()
axes.auto_scale_xyz(axes_lims, axes_lims, axes_lims)

axes.set_title('Sample Shape: Cuboid')
axes.set_xlabel('X / m')
axes.set_ylabel('Y / m')
axes.set_zlabel('Z / m')

plt.show()
```

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

StephenSmith25 commented on Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp:
```diff
@@ -5,18 +5,52 @@
+
+  auto vertices = self.getTriangleVertices();
+  auto triangles = self.getTriangleFaces();
+  size_t numberTriangles = triangles.size() / 3;
```
CSGObject has a method to find the number of triangles right?

DanielMurphy22 replied:
It does and I just tried it, but it doesn't work. I gets blocked by a canTriangulate() call (L68 in GeometryHandler.cpp) which assumes False for a CSGObject.

DanielMurphy22 replied:
I've now used `self.numberOfTriangles()`

StephenSmith25 commented on Framework/Geometry/src/Rendering/GeometryHandler.cpp:
```diff
@@ -75,17 +75,55 @@ size_t GeometryHandler::numberOfPoints() const {
 
-const std::vector<double> &GeometryHandler::getTriangleVertices() const {
+const std::vector<double> GeometryHandler::renderCuboid(const detail::ShapeInfo &shapeInfo) {
+  const auto &points = shapeInfo.points();
```
I wonder if we should implement renderCuboid and all store m_MeshPoint and m_meshVertices in the handler class. It looks like the handler is providing an interface for the rendering, rather than implementing it. Maybe this should be separated out into another class

DanielMurphy22 replied:
This could be deleted as this rendering can already be handled by OpenCascade in `RenderingHelpersOpenGL.cpp`.

StephenSmith25 reviewed (changes requested):
Instructions work and the mesh object looks correct. Just some small comments. Also I'm wondering whether the cuboid mesh stuff should be implemented in the handler class. I think it would be better to put it into its own class.

DanielMurphy22 commented on Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp:
```diff
@@ -26,5 +60,6 @@ void export_Object() {
 
-      .def("volume", &CSGObject::volume, arg("self"), "Returns the volume of this shape.");
+      .def("volume", &CSGObject::volume, arg("self"), "Returns the volume of this shape.")
+      .def("getMesh", &wrapMeshWithNDArray, (arg("self")), "Get the vertices, grouped by triangles, from mesh");
```
hmmm... so I appreciate that getMeshVertices may provide more info, but as is the case for both Mesh and CSGObjects, you use the MeshTriangles and MeshVertices to create the Mesh, that's why I simply called the result getMesh(). Whether I leave it as getMesh or change it, it needs to be identical to the method on MeshObjects: https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/mantid/geometry/src/Exports/MeshObject.cpp#L53

DanielMurphy22 commented on Framework/Geometry/inc/MantidGeometry/Rendering/GeometryHandler.h:
```diff
@@ -79,13 +83,15 @@ class MANTID_GEOMETRY_DLL GeometryHandler {
   /// get the number of points or vertices
   size_t numberOfPoints() const;
+  /// Get Mesh vertices for a CSG Cuboid
+  const std::vector<double> renderCuboid(const detail::ShapeInfo &shapeInfo);
```
I've decided to change to void as I'm editing the m_meshVertices member object anyway so I don't need to return it!

DanielMurphy22 commented on Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp:
```diff
@@ -5,18 +5,52 @@
+using Mantid::Geometry::IObject;
+// using Mantid::Geometry::MeshObject;
+using namespace Mantid::PythonInterface::Converters;
+using namespace boost::python;
```
I've trimmed down to one instance, but some of the code I didn't write already use it, so not removing both!

DanielMurphy22 commented on Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp:
```diff
@@ -5,18 +5,52 @@
+#include "MantidPythonInterface/core/Converters/WrapWithNDArray.h"
+#include "MantidPythonInterface/core/GetPointer.h"
+
+#include <boost/python/register_ptr_to_python.hpp>
```
Great point, I think I've tidied it now. I'm definitely guilty of putting in too many includes!

DanielMurphy22 commented:
This PR will need to be reworked as the scope has changed to render all CSG shapes so I'm setting it back to draft for the moment!

DanielMurphy22 committed:
Render Meshes for CSGObjects

Call triangulator directly to render with OpenCascade
This can create shapes such as spheres or cylinders
and can merge shapes

StephenSmith25 reviewed (dismissed):
Looks good! 👍

martyngigg commented on Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp:
```diff
@@ -5,18 +5,50 @@
+boost::python::object wrapMeshWithNDArray(const CSGObject *self) {
+  // PyArray_SimpleNewFromData doesn't interact well with smart pointers so use raw pointer
+
+  auto localTriangulator = new GeometryTriangulator(self);
```
I can't see a delete for this object so I think this will cause a memory leak.?Can we not just create this on the stack as `localTriangulator` is only used here?

martyngigg commented on Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp:
```diff
@@ -5,18 +5,50 @@
+  // PyArray_SimpleNewFromData doesn't interact well with smart pointers so use raw pointer
+
+  auto localTriangulator = new GeometryTriangulator(self);
+  auto vertices = localTriangulator->getTriangleVertices();
```
I think we can make all of these variables `const` and capture `vertices` and `triangles` by `const &` to avoid a copy?

DanielMurphy22 replied:
Tell me if I've now done this correctly or not!?

DanielMurphy22 replied:
This and the read-only flag were based on what I had added previously to Exports/MeshObject.cpp so I've made the const and ReadWrite updates to both.

martyngigg commented on Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp:
```diff
@@ -5,18 +5,50 @@
+    } // for each coordinate of that corner
+  }   // for each corner of the triangle
+
+  PyObject *ndarray = Impl::wrapWithNDArray(meshCoords, 3, dims, NumpyWrapMode::ReadOnly, OwnershipMode::Python);
```
As this is a copy do we need to set the read-only flag? I'm not sure it's what users would expect?

DanielMurphy22 committed:
Requested Changes

Use smart pointer
Capture variables as const &
Allow ReadWrite access to output NumpyArray
Also align thses changes on Exports/MeshObjects.cpp

martyngigg commented on Framework/PythonInterface/mantid/geometry/src/Exports/CSGObject.cpp:
```diff
@@ -5,18 +5,50 @@
+boost::python::object wrapMeshWithNDArray(const CSGObject *self) {
+  // PyArray_SimpleNewFromData doesn't interact well with smart pointers so use raw pointer
+
+  auto localTriangulator = std::make_unique<GeometryTriangulator>(self);
```
Thanks for making the changes. This is nearly there!

I think here we can simply do:
```suggestion
  GeometryTriangulator localTriangulator(self);
```

and then later use `localTriangulator.` as if it's local there is no need to heap allocate with `new/make_unique`

DanielMurphy22 replied:
Should be right now! I don't know why I didn't think of that!
